### PR TITLE
Add ability to create CC REST API users

### PR DESCRIPTION
**Pull request by kyguy**
### Type of change

- Enhancement / new feature

### Description

This PR gives ability to create REST API users for the Cruise Control REST API.
This allows developers and third-party applications to define roles and permissions to access to the Cruise Control REST API without having to disable HTTP basic authentication.

This PR adds an `apiUsers` section to the cruiseControl.spec where an API users can be defined by name and predefined Cruise Control role [1].  The schema looks like this:

```
spec:
  cruiseControl:
    apiUsers:
      - name: user
        role: USER (3)
     ...
```
Strimzi will generate a `Secret` containing a password for the specified API user.

We can also supply a custom password for the specified API user via a `Secret` like this:
```
spec:
  cruiseControl:
    apiUsers:
      - name: user
        password:
          valueFrom:
            secretKeyRef:
              name: my-secret (1)
              key: my-password (2)
        role: USER (3)
     ...
```
(1) The name of the secret containing the predefined password.
(2) The key for the password stored inside the secret.
(3) By default Cruise Control defines three roles: VIEWER, USER and ADMIN. [1]

```
VIEWER role: has access to the most lightweight kafka_cluster_state, user_tasks and review_board endpoints.
USER role: has access to all the GET endpoints except bootstrap and train.
ADMIN role: has access to all endpoints.
```
[1] https://github.com/linkedin/cruise-control/wiki/Security#authorization

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc [hunk 1]
```diff
@@ -147,6 +147,102 @@ spec:
...
+            secretKeyRef:
+              name: cruise-control-api-user-secret (2)
+              key: password (3)
+        role: USER (3)
+# ...
+----
+1. The name of the API user.
+2. The name of the secret containing the predefined password.
+3. The key for the password stored inside the secret.
+4. Valid API roles are VIEWER, USER, and ADMIN".
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:244]:
```suggestion
<1> The name of the API user.
<2> The name of the secret containing the predefined password.
<3> The key for the password stored inside the secret.
<4> Valid API roles are `VIEWER`, `USER`, and `ADMIN`.
```

File: documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc [hunk 2]
```diff
@@ -147,6 +147,102 @@ spec:
...
+  cruiseControl:
+    apiUsers:
+      - name: my-user (1)
+        role: USER (2)
+# ...
+----
+1. The name of the API user.
+2. Valid API roles are VIEWER, USER, and ADMIN".
+For more information on valid API user roles see https://github.com/linkedin/cruise-control/wiki/Security#authorization
+
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:170]:
```suggestion

```

Comment (PaulRMellor) [reply documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:170]: We have this link in two places. Maybe we can remove them and put the link in a note as below.

File: documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc [hunk 3]
```diff
@@ -147,6 +147,102 @@ spec:
...
+
+When an API user is created, Strimzi generates a random password for that user.
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:172]:
```suggestion
NOTE: For more information on valid API user roles, see the link:https://github.com/linkedin/cruise-control/wiki/Security#authorization[Cruise Control documentation^].

When an API user is created, Strimzi generates a random password for that user.
```

File: documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc [hunk 4]
```diff
@@ -147,6 +147,102 @@ spec:
...
+              name: cruise-control-api-user-secret (2)
+              key: password (3)
+        role: USER (3)
+# ...
+----
+1. The name of the API user.
+2. The name of the secret containing the predefined password.
+3. The key for the password stored inside the secret.
+4. Valid API roles are VIEWER, USER, and ADMIN".
+For more information on valid API user roles see https://github.com/linkedin/cruise-control/wiki/Security#authorization
```
Comment (PaulRMellor) [inline documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:245]:
```suggestion

```

Comment (PaulRMellor) [reply documentation/api/io.strimzi.api.kafka.model.CruiseControlSpec.adoc:245]: If we go with the note suggestion

File: api/src/main/java/io/strimzi/api/kafka/model/balancing/ApiUser.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    @Description("Specify the password for the user Cruise Control REST API user.")
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/balancing/ApiUser.java:59]:
```suggestion
    @Description("The password for the Cruise Control REST API user.")
```

File: api/src/main/java/io/strimzi/api/kafka/model/balancing/ApiUser.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public void setPassword(Password password) {
+        this.password = password;
+    }
+
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    @Pattern("^(ADMIN)|(USER)|(VIEWER)$")
+    @Description("Cruise Control REST API user role." +
+            "Valid API user roles are VIEWER, USER, and ADMIN" +
```
Comment (PaulRMellor) [inline api/src/main/java/io/strimzi/api/kafka/model/balancing/ApiUser.java:71]:
```suggestion
            "Valid API user roles are VIEWER, USER, and ADMIN." +
```

File: documentation/modules/appendix_crds.adoc [hunk 7]
```diff
@@ -1453,6 +1455,51 @@ Used in: xref:type-BrokerCapacity-{context}[`BrokerCapacity`]
...
+
+
+[options="header"]
+|====
+|Property         |Description
+|name      1.2+<.<a|Cruise Control REST API user. The following names are reserved for Strimzi and cannot be used: admin, healthcheck.
+|string
+|password  1.2+<.<a|Specify the password for the user Cruise Control REST API user.
+|xref:type-Password-{context}[`Password`]
+|role      1.2+<.<a|Cruise Control REST API user role.Valid API user roles are VIEWER, USER, and ADMINFor more information on valid API user roles see https://github.com/linkedin/cruise-control/wiki/Security#authorization.
```
Comment (PaulRMellor) [inline documentation/modules/appendix_crds.adoc:1471]: Spaces between the sentences? Do we need the link here? Certainly in the docs, we don't need it as we reference it earlier in the section. Maybe it's good to have it in the yaml etc.

**Review by PaulRMellor — COMMENTED** [review]
Looks great Kyle. 
I've left some suggestions.
Let me know if anything's unclear

Maybe we'll need something in the deploying guide, which we can talk about.

**Review by fvaleri — COMMENTED** [review]
@kyguy nice work. Left some comments.

Looks like the `my-cluster-cruise-control-api` secret stores the user password in base64 form and you need to encode your password before using it. I'm not sure this is the intended behavior. If yes, then we should make it clear in the documentation.

```sh
$ echo "Y2hhbmdlaXQ=" | base64 -d
changeit

$ kubectl exec -it my-cluster-cruise-control-59699c88bf-24ft8 -- curl --cacert /etc/cruise-control/cc-certs/cruise-control.crt -u "foo:Y2hhbmdlaXQ=" https://localhost:9090/kafkacruisecontrol/state
MonitorState: {state: RUNNING(19.200% trained), NumValidWindows: (1/1) (100.000%), NumValidPartitions: 117/117 (100.000%), flawedPartitions: 0}
ExecutorState: {state: NO_TASK_IN_PROGRESS}
AnalyzerState: {isProposalReady: true, readyGoals: [NetworkInboundUsageDistributionGoal, PreferredLeaderElectionGoal, CpuUsageDistributionGoal, PotentialNwOutGoal, LeaderReplicaDistributionGoal, NetworkInboundCapacityGoal, LeaderBytesInDistributionGoal, DiskCapacityGoal, ReplicaDistributionGoal, RackAwareGoal, MinTopicLeadersPerBrokerGoal, TopicReplicaDistributionGoal, NetworkOutboundCapacityGoal, CpuCapacityGoal, DiskUsageDistributionGoal, NetworkOutboundUsageDistributionGoal, ReplicaCapacityGoal]}
AnomalyDetectorState: {selfHealingEnabled:[], selfHealingDisabled:[DISK_FAILURE, BROKER_FAILURE, GOAL_VIOLATION, METRIC_ANOMALY, TOPIC_ANOMALY, MAINTENANCE_EVENT], selfHealingEnabledRatio:{DISK_FAILURE=0.0, BROKER_FAILURE=0.0, GOAL_VIOLATION=0.0, METRIC_ANOMALY=0.0, TOPIC_ANOMALY=0.0, MAINTENANCE_EVENT=0.0}, recentGoalViolations:[], recentBrokerFailures:[], recentMetricAnomalies:[], recentDiskFailures:[], recentTopicAnomalies:[], recentMaintenanceEvents:[], metrics:{meanTimeBetweenAnomalies:{GOAL_VIOLATION:0.00 milliseconds, BROKER_FAILURE:0.00 milliseconds, METRIC_ANOMALY:0.00 milliseconds, DISK_FAILURE:0.00 milliseconds, TOPIC_ANOMALY:0.00 milliseconds}, meanTimeToStartFix:0.00 milliseconds, numSelfHealingStarted:0, numSelfHealingFailedToStart:0, ongoingAnomalyDuration=0.00 milliseconds}, ongoingSelfHealingAnomaly:None, balancednessScore:100.000}
```

Another problem I found is that, after the initial deploy, new users are not added to `my-cluster-cruise-control-api` secret, even after forcing a pod restart.

It looks like I'm not able to run the ST (CC goes into crash loop). Instead, it works fine with a manual deploy, so I'll try again tomorrow.

Few additional questions:

- Should we document how to enable authN and authZ logs?
- Should we add an example with a custom cc-api-user-secret?

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. I left some comments. But TBH, I think this should start with a proposal:
* This is very significant change to the Strimzi API
* It seems to be very complex for what seems like a very niche feature.
* It is a significant departure from how we do user management elsewhere (=> the `KafkaUser` resources) and possibly sets a precedent for other users (e.g. Kafka Connect). We do not want to have different solution everywhere. So I think this needs to be carefully discussed.
* As it touches security, we need to prepared to implement additional requests -> configurable password rules, password renewals, etc.

So I think this should be properly discussed in a proposal and in the full context.

----

For example ... have you considered some simplification such as having user manage a single secret with the user configuration in the Cruise Control format? That would significantly simplify the design and shift a lot of responsibilities to the user.

**Comment by ShubhamRwt** [comment]
@kyguy Hi, Going through the above comments, Is this PR now going to have a proposal first or should I review it as of now?

**Comment by kyguy** [comment]
@ShubhamRwt Going to throw a proposal together first and then make changes to this as the proposal is reviewed. Feel free to add comments/thoughts here or on the proposal

**Comment by scholzj** [comment]
@kyguy Should we close this? I think the proposal went in a different direction, so I guess you will open a brand new PR for it, or?

**Comment by kyguy** [comment]
Closing in favor of a new PR which follows direction taken by the proposal here [1]

[1] https://github.com/strimzi/proposals/pull/86

**Pull request closed without merging**